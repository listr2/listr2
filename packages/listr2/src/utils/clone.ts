import rfdc from 'rfdc'

const clone = rfdc({ circles: true })

/**
 * Deep clones a object in the easiest manner.
 */
// TODO: swap this with structured clone whenever we migrate to node 18
export function cloneObject<T extends Record<PropertyKey, any>>(obj: T): T {
  return clone(obj)
}
