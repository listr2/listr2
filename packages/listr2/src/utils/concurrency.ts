// adapted from https://github.com/fabiospampinato/promise-concurrency-limiter
// with MIT license, since it is again ESM only
// we needed to stop on the first failing case unlike the original
export class Concurrency {
  promise: Promise<void>

  private resolve: () => void
  private reject: (error: Error) => void
  private concurrency: number
  private count: number
  private queue: Set<() => void>

  constructor (options: { concurrency: number }) {
    this.concurrency = options.concurrency
    this.count = 0
    this.queue = new Set()
    this.initPromise()
  }

  public add<T>(fn: () => Promise<T>): Promise<T> {
    if (this.count < this.concurrency) {
      return this.run(fn)
    }

    return new Promise<T>((resolve) => {
      const callback = (): void => resolve(this.run(fn))

      this.queue.add(callback)
    })
  }

  private initPromise (): void {
    const { promise, resolve, reject } = this.withResolvers<void>()

    this.promise = promise
    this.resolve = resolve
    this.reject = reject
  }

  private flush (): void {
    for (const callback of this.queue) {
      if (this.count >= this.concurrency) {
        break
      }

      this.queue.delete(callback)

      callback()
    }
  }

  private run<T>(fn: () => Promise<T>): Promise<T> {
    this.count++

    const promise = fn()

    const cleanup = (): void => {
      this.count--

      this.flush()

      if (this.count === 0) {
        this.resolve()
      }
    }

    promise.then(cleanup, (error) => {
      this.queue.clear()
      this.reject(error)
    })

    return promise
  }

  private withResolvers<T>(): { promise: Promise<T>, resolve: (value: T) => void, reject: (error: Error) => void } {
    let resolve: (value: T) => void
    let reject: (error: Error) => void

    const promise = new Promise<T>((res, rej) => {
      resolve = res
      reject = rej
    })

    return {
      promise,
      resolve: resolve!,
      reject: reject!
    }
  }
}
