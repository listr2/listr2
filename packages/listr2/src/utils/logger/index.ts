export * from './logger.constants'
export type * from './logger.interface'
export * from './logger'
