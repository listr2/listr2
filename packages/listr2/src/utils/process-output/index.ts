export type * from './process-output-buffer.interface'
export * from './process-output-buffer'
export * from './process-output-stream'
export * from './process-output'
export type * from './process-output.interface'
export * from './writable'
