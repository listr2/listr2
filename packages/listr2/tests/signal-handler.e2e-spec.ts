import type { MockProcessOutput } from './utils'
import { expectProcessOutputToMatchSnapshot, mockProcessOutput, unmockProcessOutput } from './utils'
import { Listr, delay } from '@root'

describe('signal handlers', () => {
  const output: MockProcessOutput = {} as MockProcessOutput

  process.stdout.isTTY = true

  beforeEach(async() => {
    mockProcessOutput(output)
  })

  afterEach(async() => {
    unmockProcessOutput(output)
    jest.clearAllMocks()
  })

  // LFRhPFJHChAJcQIEadkDdOlPezdltcfe
  it('should try to cancel running tasks', async() => {
    await new Listr(
      [
        {
          title: 'a task',
          task: async(): Promise<void> => {
            await delay(10)

            process.emit('SIGINT')

            await delay(100)
          }
        }
      ],
      {
        concurrent: false,
        renderer: 'test',
        registerSignalListeners: true
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'LFRhPFJHChAJcQIEadkDdOlPezdltcfe')
  })

  // https://github.com/listr2/listr2/issues/709 running this double seems to mess it somehow?
  // ACxcRBILlHsqiSxYArjASFdofVqLaUyE
  it('should try to cancel running tasks', async() => {
    await new Listr(
      [
        {
          title: 'a task',
          task: async(): Promise<void> => {}
        }
      ],
      {
        concurrent: false,
        renderer: 'test',
        registerSignalListeners: true
      }
    ).run()

    await new Listr(
      [
        {
          title: 'a task',
          task: async(): Promise<void> => {
            await delay(10)

            process.emit('SIGINT')

            await delay(100)
          }
        }
      ],
      {
        concurrent: false,
        renderer: 'test',
        registerSignalListeners: true
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'ACxcRBILlHsqiSxYArjASFdofVqLaUyE')
  })

  it('should not clear other SIGINT listeners', async() => {
    process.addListener('SIGINT', function doNotRemove() {})
    expect(process.listeners('SIGINT').map((listener) => listener.name)).toContain('doNotRemove')

    await new Listr(
      [
        {
          title: 'a task',
          task: async(): Promise<void> => {}
        }
      ],
      {
        concurrent: false,
        renderer: 'test',
        registerSignalListeners: true
      }
    ).run()

    expect(process.listeners('SIGINT').map((listener) => listener.name)).toContain('doNotRemove')
  })

  it('should remove the signal handler if checking the task fails', async() => {
    const listeners = process.listeners('SIGINT')

    await new Listr([
      {
        title: 'a task',
        task: async(): Promise<void> => {},
        enabled: (): boolean => {
          throw new Error('failed')
        }
      }
    ])
      .run()
      .catch(() => {})

    expect(process.listeners('SIGINT')).toEqual(listeners)
  })
})
