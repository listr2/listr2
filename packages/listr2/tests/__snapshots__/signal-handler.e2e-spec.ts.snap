// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`signal handlers should try to cancel running tasks: ACxcRBILlHsqiSxYArjASFdofVqLaUyE-exit 1`] = `
[
  127,
]
`;

exports[`signal handlers should try to cancel running tasks: ACxcRBILlHsqiSxYArjASFdofVqLaUyE-stderr 1`] = `[]`;

exports[`signal handlers should try to cancel running tasks: ACxcRBILlHsqiSxYArjASFdofVqLaUyE-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"a task","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["a task"]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"a task","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["a task"]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"a task","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["a task"]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"a task","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":["a task"]}}
",
]
`;

exports[`signal handlers should try to cancel running tasks: LFRhPFJHChAJcQIEadkDdOlPezdltcfe-exit 1`] = `
[
  127,
]
`;

exports[`signal handlers should try to cancel running tasks: LFRhPFJHChAJcQIEadkDdOlPezdltcfe-stderr 1`] = `[]`;

exports[`signal handlers should try to cancel running tasks: LFRhPFJHChAJcQIEadkDdOlPezdltcfe-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"a task","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["a task"]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"a task","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":["a task"]}}
",
]
`;
